marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs

```diff
@@ -254,27 +262,104 @@ scriptTxWorker onInsert coordinator path = do
+  -> FilePath
+  -> IO (IO b, MVar EpochSPDIndex)
+epochStakepoolSizeWorker_
+        nodeConfigPath
+        onInsert
+        securityParam
+        Coordinator{_barrier}
+        ch
+        dbPath = do
+    let ledgerStateDir = takeDirectory dbPath </> "ledgerStates"
+    createDirectoryIfMissing False ledgerStateDir
+    indexerMVar <- newMVar =<< EpochSPD.open dbPath ledgerStateDir securityParam
+
+    nodeConfigE <- runExceptT $ readNetworkConfig (NetworkConfigFile nodeConfigPath)
+    nodeConfig <- either (error . show) pure nodeConfigE
+    genesisConfigE <- runExceptT $ readCardanoGenesisConfig nodeConfig
+    genesisConfig <- either (error . show . renderGenesisConfigError) pure genesisConfigE
+
+    let initialLedgerState = O.ledgerState $ initExtLedgerStateVar genesisConfig
+        hfLedgerConfig = O.topLevelConfigLedger $ O.pInfoConfig (mkProtocolInfoCardano genesisConfig)
+
+        loop currentLedgerState currentEpochNo previousEpochNo = do
+            signalQSemN _barrier 1
+            chainSyncEvent <- atomically $ readTChan ch
+
+            newLedgerState <- case chainSyncEvent of
+              RollForward blockInMode@(C.BlockInMode (C.Block (C.BlockHeader slotNo bh bn) _) _) chainTip -> do
+                  -- If the block is rollbackable, we always store the LedgerState. If the block is
+                  -- immutable, we only store it right at the beginning of a new epoch.
+                  let isFirstEventOfEpoch = currentEpochNo > previousEpochNo
+                  let storableEvent =
+                          EpochSPD.toStorableEvent
+                            currentLedgerState
+                            slotNo
+                            bh
+                            bn
+                            chainTip
+                            securityParam
+                            isFirstEventOfEpoch
+                  newIndexer <-
+                      modifyMVar indexerMVar
+                        $ \idx -> fmap (\newIdx -> (newIdx, newIdx)) $ Storable.insert storableEvent idx
+                  onInsert (newIndexer, storableEvent)
+
+                  -- Compute new LedgerState given block and old LedgerState
+                  pure $ O.lrResult
+                       $ O.tickThenReapplyLedgerResult
+                            hfLedgerConfig
+                            (C.toConsensusBlock blockInMode)
+                            currentLedgerState
+
+              RollBackward C.ChainPointAtGenesis _ct -> do
+                  modifyMVar_ indexerMVar $ \ix -> fromMaybe ix <$> Storable.rewind C.ChainPointAtGenesis ix
+                  pure initialLedgerState
+              RollBackward cp _ct ->
+                  modifyMVar indexerMVar $ \ix -> do
+                  newIndex <- fromMaybe ix <$> Storable.rewind cp ix
+                  -- The possible points from which we can possibly rollback should be available
+                  -- in the buffer events and from the resumable points.
```

Comment:
If we store only the first ledger state from epoch on disk but the rollback doesn't go to exactly that chain point then I think we actually need to roll back *to that first slot of the epoch* for which we have ledger state for, and then also reconnect local chainsync and start reading blocks from that exact slot.

Reply:
Oops, out of date comment. Sorry :(. See change: 493f6797e65c571592f19db97d6f069b625c8116